src/form_coder.gleam

```diff
@@ -1,50 +1,86 @@
-import gleam/dynamic.{Dynamic}
+//// x-www-form-urlencoded codec
+////
+//// The codec follows the WHATWG specification available at:
+//// https://url.spec.whatwg.org/#application/x-www-form-urlencoded but uses
+//// UTF-8 as the only supported encoding.
+
+import gleam/dynamic
 import gleam/list
 import gleam/map.{Map}
 import gleam/uri
 import gleam/result
 import gleam/string
+import gleam/string_builder
+import gleam/option
 
 pub type Query {
   QStr(String)
   QList(List(Query))
   QMap(Map(String, Query))
 }
 
+/// Characters not encoded by Gleam's percent encoding, but that we need to encode
+const not_encoded_by_gleam = [
+  #("!", "%21"),
+  #("$", "%24"),
+  #("'", "%27"),
+  #("(", "%28"),
+  #(")", "%29"),
+  #("+", "%2B"),
+  #("~", "%7E"),
+]
```

Comment:
Curious, do we know why these are not percent encoded by Gleam's `uri`?

Reply:
There's varying targets for percent encoding, so what you want to encode depends on the context you are encoding to. Gleam follows `encodeURIComponent` in JS IIRC, and an almost similar set of characters in Erlang.